toolchain/lex/lex.cpp

```diff
@@ -1179,17 +1175,43 @@ auto Lexer::LexWordAsTypeLiteralToken(llvm::StringRef word, int32_t byte_offset)
       return LexResult::NoMatch();
   };
 
+  // No leading zeros allowed.
+  if ('1' > word[1] || word[1] > '9') {
+    return LexResult::NoMatch();
+  }
+
   llvm::StringRef suffix = word.substr(1);
-  if (!CanLexInt(emitter_, suffix)) {
+  constexpr ssize_t DigitLimit = std::numeric_limits<int64_t>::digits10;
+  if (suffix.size() > DigitLimit) {
+    // See if this is not actually a type literal.
+    if (!llvm::all_of(suffix, IsDecimalDigit)) {
+      return LexResult::NoMatch();
+    }
+
+    // Otherwise, diagnose and produce an error token.
+    CARBON_DIAGNOSTIC(TooManyTypeBitWidthDigits, Error,
+                      "found a type literal with a bit width using {0} digits, "
+                      "which is greater than the limit of {1}",
+                      size_t, size_t);
+    emitter_.Emit(word.begin() + 1, TooManyTypeBitWidthDigits, suffix.size(),
+                  DigitLimit);
     return LexTokenWithPayload(TokenKind::Error, word.size(), byte_offset);
   }
-  llvm::APInt suffix_value;
-  if (suffix.getAsInteger(10, suffix_value)) {
-    return LexResult::NoMatch();
+
+  // It's tempting to do something more clever because we know the length ahead
+  // of time, but we expect these to be short (1-3 digits) and profiling doesn't
+  // show the loop as hot in the short cases.
+  ssize_t suffix_value = suffix[0] - '0';
+  for (char c : suffix.drop_front()) {
+    if (!IsDecimalDigit(c)) {
+      return LexResult::NoMatch();
+    }
+    suffix_value = suffix_value * 10 + (c - '0');
```

Comment:
I'm trying to see if there's some pre-existing implementation we could use, maybe `std::from_chars(suffix.begin(), suffix.end(), suffix_value);`?

```suggestion
  // It's tempting to do something more clever because we know the length ahead
  // of time, but we expect these to be short (1-3 digits) and profiling doesn't
  // show the loop as hot in the short cases.
  if (std::from_chars(suffix.begin(), suffix.end(), suffix_value) != suffix.end()) {
    return LexResult::NoMatch();
```

Reply:
The API here is a bit awkward, but found a way to make it work.